dart-mode.el

```diff
@@ -562,6 +579,251 @@ This uses `dart-format-path' to find the formatter."
+  (if (not dart-analysis-server) (dart-start-analysis-server))
+  ;; TODO(hterkelsen): Add this file to the priority files.
+  (dart-add-current-file-to-analysis-roots)
+  (dart-add-analysis-server-change-hooks)
+  (add-to-list 'flycheck-checkers 'dart-analysis-server))
+
+(defun dart-start-analysis-server ()
+  "Start the Dart analysis server.
+
+`dart' must be on the PATH."
+  (let ((dart-process
+         (start-process "dart-analysis-server"  "*dart-analysis-server*"
+                        "dart" dart-analysis-server-snapshot-path
+                        "--no-error-notification")))
+    (setq dart-analysis-server (dart-analysis-server-create
+                                dart-process))))
+
+(defun dart-analysis-server-create (process)
+  "Create a Dart analysis server from PROCESS."
+  (let* ((das-buffer (generate-new-buffer (process-name process)))
+         (das-inst (make-dart-analysis-server
+                    :process process
+                    :buffer das-buffer
+                    :queue nil)))
+    (buffer-disable-undo (dart-analysis-server-buffer das-inst))
+    (set-process-filter
+     process
+     `(lambda (proc string)
+        (dart-analysis-server-process-filter ',das-inst string)))
+    das-inst))
+
+(defun dart-add-analysis-server-change-hooks ()
+  "Report the dirty buffer state to Dart analysis server."
+  (add-hook 'first-change-hook 'dart-add-analysis-overlay t t)
+  (add-hook 'after-change-functions 'dart-change-analysis-overlay t t)
+  (add-hook 'after-save-hook 'dart-remove-analysis-overlay t t))
+
+(defun dart-add-analysis-overlay ()
+  "Report to Dart analysis server that it should overlay this buffer."
+  (dart-analysis-server-send
+   "analysis.updateContent"
+   `((files .
+            ((,buffer-file-name . ((type . "add")
+                                   (content . ,(buffer-string)))))))))
+
+(defun dart-change-analysis-overlay
+    (change-begin change-end change-before-length)
+  "Report to analysis server that it should change the overlay for this buffer.
+
+The region that changed ranges from CHANGE-BEGIN to CHANGE-END, and the
+length of the text before the change is CHANGE-BEFORE-LENGTH."
+  (combine-after-change-calls
+    (dart-analysis-server-send
+     "analysis.updateContent"
+     `((files
+        . ((,buffer-file-name
+            . ((type . "change")
+               (edits
+                . (((offset . ,change-begin)
+                    (length . ,change-before-length)
+                    (replacement
+                     . ,(buffer-substring change-begin change-end)))))))))))))
+
+(defun dart-remove-analysis-overlay ()
+  "Remove the overlay for the current buffer since it has been saved."
+  (dart-analysis-server-send
+   "analysis.updateContent"
+   `((files . ((,buffer-file-name . ((type . "remove"))))))))
+
+(defun dart-add-current-file-to-analysis-roots ()
+  "Add the current file's pub root or directory to analysis roots."
```

Comment:
Nit: "analysis roots" -> "the analysis server's analysis roots".

Same thing below.